tests/utils/test_matrix_operations.py

```diff
@@ -144,60 +95,254 @@ def test_csc_slice():
+    slicer = matrix_operations.MatrixSlicer(
+        domain_inds, range_inds, range_size, domain_size
+    )
+
+    if transpose:
+        # First transpose the slicer.
+        slicer = slicer.T
+        # Now, the number of rows in the resulting matrix is determined from domain
+        # information, since the slicer is transposed.
+        if domain_size is not None:
+            # If the domain size is given, the number of rows is given by this.
+            num_rows = domain_size
+        elif domain_inds is not None:
+            # If no domain size is given but domain indices are, the output number of
+            # rows is the maximum index + 1 (because indices are 0-offset).
+            num_rows = domain_inds.max() + 1
+        else:
+            # If no domain size or indices are given, the output number of rows is the
+            # number of range indices.
+            num_rows = range_inds.size
+    else:
+        # Get the number of output rows from the range information. The logic is the
+        # same as for the transpose case.
+        if range_size is not None:
+            num_rows = range_size
+        elif range_inds is not None:
+            num_rows = range_inds.max() + 1
+        else:
+            num_rows = domain_inds.size
+
+    if range_inds is None:
+        range_inds = np.arange(domain_inds.size)
+    if domain_inds is None:
+        domain_inds = np.arange(range_inds.size)
+
+    result = slicer @ target
+
+    if mode == "dense":
+        # No special handling of transpose for vectors.
+        A_known = np.zeros(num_rows)
+        if transpose:
+            A_known[domain_inds] = target[range_inds]
+        else:
+            A_known[range_inds] = target[domain_inds]
+        assert np.allclose(result, A_known)
+    elif mode == "sparse":
+        A_known = np.zeros((num_rows, target.shape[1]))
+        if transpose:
+            A_known[domain_inds] = target.toarray()[range_inds]
+        else:
+            A_known[range_inds] = target.toarray()[domain_inds]
+        assert np.allclose(result.toarray(), A_known)
+    else:
+        val_known = np.zeros(num_rows)
+        if transpose:
+            val_known[domain_inds] = target.val[range_inds]
+        else:
+            val_known[range_inds] = target.val[domain_inds]
+        assert np.allclose(result.val, val_known)
+        jac_known = np.zeros((num_rows, target.jac.shape[1]))
+        if transpose:
+            jac_known[domain_inds] = target.jac.toarray()[range_inds]
+        else:
+            jac_known[range_inds] = target.jac.toarray()[domain_inds]
+        assert np.allclose(result.jac.toarray(), jac_known)
+
+
+@pytest.mark.parametrize(
+    "other_mode, target_mode, operator",
+    [
+        # The following tuples are in the form (A, B, operator), where A and B are the
+        # types of the operands and operator is the operator to be applied. The cases
+        # considered span all permissible combinations of operands and operators. Note
```

Comment:
Should we consider doing a check that all combinations are covered to enforce updates of this test if new combinations are introduced?

Reply:
In principle, yes. In practice, do you have ideas for how to implement this? 

It is not clear to me what such a new case would be, but you are right that  if it shows up at a later point, we may not remember to test it.

Reply:
Update: Having just added a new case, essentially broadcasting of Ad scalars, I agree even more that such an enforcement is warranted. Still, I am blank on how to do this. Suggestions? If none, We will just have to go for disciplin.

Reply:
Update: The new set of tests now covers all relevant operators and operands. I decided not to check that errors are raised if an unpermissible combination is tried because: 1) The most likely such combinations involve numpy as the 'other operand', and numpy will parse this to a meaningless expression no matter what we do; 2) Other cases can basically involve anything, and we can't check for that. In all likelihood, an error of a sort will be raised at some point, and the user will just have to debug from there.